components/layout/src/lumino/common.ts

```diff
@@ -0,0 +1,56 @@
+import { isTrue, HPCCResizeElement } from "@hpcc-js/wc-core";
+import { Widget } from "@lumino/widgets";
+import { WidgetAdapter } from "./widgetAdapter";
+
+let hpccSlotID = 0;
+
+export abstract class HPCCLuminoElement extends HPCCResizeElement {
+
+    protected _slot: HTMLSlotElement;
+    private _childCount = 0;
+
+    constructor() {
+        super();
+        this.createPanel();
+        this.construct();
+        this._slot.addEventListener("slotchange", () => this.construct());
+    }
+
+    abstract createPanel(): void;
+    abstract addWidget(w: WidgetAdapter, _e: HTMLElement, _ref?: Widget): void;
+
+    private construct() {
+        const codeElements = this._slot.assignedElements();
+        const widgetIdx: { [id: string]: WidgetAdapter } = {};
+        for (let i = 0; i < codeElements.length; ++i) {
+            ++this._childCount;
+            const e = codeElements[i] as HTMLElement;
+            const slot = document.createElement("slot");
+            slot.setAttribute("name", `slot_${++hpccSlotID}`);
+            slot.style.cssText = e.style.cssText;
+            slot.style.display = "inline-block";
+            const w = new WidgetAdapter(this, slot);
+            widgetIdx[e.id] = w;
+            w.title.label = e.dataset.label || (e.id && `#${e.id}`) || `${e.tagName} ${this._childCount}`;
```

Comment:
Do any of these need ?. or anything? I didn't see anything about how / whether dataset was getting defined

Reply:
No `e.dataset` is always there for a HTMLElement (see https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/dataset) - however it may be empty.

Reply:
Oh, data- attributes... Right, sorry. I had ecl on the brain.